File: ios/Classes/FlutterMediaPlayerPlugin.swift
import Flutter
import UIKit
import AVFoundation

public class FlutterMediaPlayerPlugin: NSObject, FlutterPlugin {
    private var playerManager: MediaPlayerManager!
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutter_media_player", binaryMessenger: registrar.messenger())
        let instance = FlutterMediaPlayerPlugin()
        
        // Initialize player manager
        instance.playerManager = MediaPlayerManager(methodChannel: channel)
        
        registrar.addMethodCallDelegate(instance, channel: channel)
        
        // Register platform view factory
        registrar.register(
            MediaPlayerViewFactory(playerManager: instance.playerManager),
            withId: "flutter_media_player_view"
        )
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            handleInitialize(call, result: result)
        case "load":
            handleLoad(call, result: result)
        case "setPlaylist":
            handleSetPlaylist(call, result: result)
        case "play":
            handlePlay(call, result: result)
        case "pause":
            handlePause(call, result: result)
        case "stop":
            handleStop(call, result: result)
        case "seekTo":
            handleSeekTo(call, result: result)
        case "setVolume":
            handleSetVolume(call, result: result)
        case "setSpeed":
            handleSetSpeed(call, result: result)
        case "setMuted":
            handleSetMuted(call, result: result)
        case "setBoxFit":
            handleSetBoxFit(call, result: result)
        case "setSubtitleTrack":
            handleSetSubtitleTrack(call, result: result)
        case "skipToIndex":
            handleSkipToIndex(call, result: result)
        case "updateConfig":
            handleUpdateConfig(call, result: result)
        case "dispose":
            handleDispose(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func handleInitialize(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID is required", details: nil))
            return
        }
        
        let config = args["config"] as? [String: Any]
        
        do {
            try playerManager.initializePlayer(playerId: playerId, config: config)
            result(nil)
        } catch {
            result(FlutterError(code: "INITIALIZATION_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleLoad(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String,
              let mediaItem = args["mediaItem"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID and media item are required", details: nil))
            return
        }
        
        do {
            try playerManager.loadMediaItem(playerId: playerId, mediaItem: mediaItem)
            result(nil)
        } catch {
            result(FlutterError(code: "LOAD_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleSetPlaylist(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String,
              let playlist = args["playlist"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID and playlist are required", details: nil))
            return
        }
        
        let startIndex = args["startIndex"] as? Int ?? 0
        
        do {
            try playerManager.setPlaylist(playerId: playerId, playlist: playlist, startIndex: startIndex)
            result(nil)
        } catch {
            result(FlutterError(code: "PLAYLIST_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handlePlay(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID is required", details: nil))
            return
        }
        
        do {
            try playerManager.play(playerId: playerId)
            result(nil)
        } catch {
            result(FlutterError(code: "PLAY_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handlePause(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID is required", details: nil))
            return
        }
        
        do {
            try playerManager.pause(playerId: playerId)
            result(nil)
        } catch {
            result(FlutterError(code: "PAUSE_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleStop(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID is required", details: nil))
            return
        }
        
        do {
            try playerManager.stop(playerId: playerId)
            result(nil)
        } catch {
            result(FlutterError(code: "STOP_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleSeekTo(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String,
              let position = args["position"] as? Int64 else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID and position are required", details: nil))
            return
        }
        
        do {
            try playerManager.seekTo(playerId: playerId, position: position)
            result(nil)
        } catch {
            result(FlutterError(code: "SEEK_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleSetVolume(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String,
              let volume = args["volume"] as? Double else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID and volume are required", details: nil))
            return
        }
        
        do {
            try playerManager.setVolume(playerId: playerId, volume: Float(volume))
            result(nil)
        } catch {
            result(FlutterError(code: "VOLUME_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleSetSpeed(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String,
              let speed = args["speed"] as? Double else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID and speed are required", details: nil))
            return
        }
        
        do {
            try playerManager.setPlaybackSpeed(playerId: playerId, speed: Float(speed))
            result(nil)
        } catch {
            result(FlutterError(code: "SPEED_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleSetMuted(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String,
              let muted = args["muted"] as? Bool else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID and muted state are required", details: nil))
            return
        }
        
        do {
            try playerManager.setMuted(playerId: playerId, muted: muted)
            result(nil)
        } catch {
            result(FlutterError(code: "MUTE_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleSetBoxFit(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String,
              let boxFit = args["boxFit"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID and box fit are required", details: nil))
            return
        }
        
        do {
            try playerManager.setBoxFit(playerId: playerId, boxFit: boxFit)
            result(nil)
        } catch {
            result(FlutterError(code: "BOXFIT_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleSetSubtitleTrack(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID is required", details: nil))
            return
        }
        
        let subtitleTrack = args["subtitleTrack"] as? [String: Any]
        
        do {
            try playerManager.setSubtitleTrack(playerId: playerId, subtitleTrack: subtitleTrack)
            result(nil)
        } catch {
            result(FlutterError(code: "SUBTITLE_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleSkipToIndex(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String,
              let index = args["index"] as? Int else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID and index are required", details: nil))
            return
        }
        
        do {
            try playerManager.skipToIndex(playerId: playerId, index: index)
            result(nil)
        } catch {
            result(FlutterError(code: "SKIP_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleUpdateConfig(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String,
              let config = args["config"] as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID and config are required", details: nil))
            return
        }
        
        do {
            try playerManager.updateConfig(playerId: playerId, config: config)
            result(nil)
        } catch {
            result(FlutterError(code: "CONFIG_ERROR", message: error.localizedDescription, details: nil))
        }
    }
    
    private func handleDispose(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let playerId = args["playerId"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENT", message: "Player ID is required", details: nil))
            return
        }
        
        do {
            try playerManager.disposePlayer(playerId: playerId)
            result(nil)
        } catch {
            result(FlutterError(code: "DISPOSE_ERROR", message: error.localizedDescription, details: nil))
        }
    }
}

File: android/src/main/kotlin/com/example/flutter_media_player/FlutterMediaPlayerPlugin.kt
package com.example.flutter_media_player

import android.content.Context
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.platform.PlatformViewRegistry

/**
 * FlutterMediaPlayerPlugin
 */
class FlutterMediaPlayerPlugin: FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context
    private lateinit var playerManager: MediaPlayerManager

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_media_player")
        channel.setMethodCallHandler(this)
        
        // Initialize player manager
        playerManager = MediaPlayerManager(context, channel)
        
        // Register platform view
        flutterPluginBinding
            .platformViewRegistry
            .registerViewFactory(
                "flutter_media_player_view",
                MediaPlayerViewFactory(playerManager)
            )
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "initialize" -> handleInitialize(call, result)
            "load" -> handleLoad(call, result)
            "setPlaylist" -> handleSetPlaylist(call, result)
            "play" -> handlePlay(call, result)
            "pause" -> handlePause(call, result)
            "stop" -> handleStop(call, result)
            "seekTo" -> handleSeekTo(call, result)
            "setVolume" -> handleSetVolume(call, result)
            "setSpeed" -> handleSetSpeed(call, result)
            "setMuted" -> handleSetMuted(call, result)
            "setBoxFit" -> handleSetBoxFit(call, result)
            "setSubtitleTrack" -> handleSetSubtitleTrack(call, result)
            "setQualityTrack" -> handleSetQualityTrack(call, result)
            "setAudioTrack" -> handleSetAudioTrack(call, result)
            "enableAutoQuality" -> handleEnableAutoQuality(call, result)
            "skipToIndex" -> handleSkipToIndex(call, result)
            "updateConfig" -> handleUpdateConfig(call, result)
            "dispose" -> handleDispose(call, result)
            else -> result.notImplemented()
        }
    }

    private fun handleInitialize(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val config = call.argument<Map<String, Any>>("config")
            
            if (playerId != null) {
                playerManager.initializePlayer(playerId, config)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID is required", null)
            }
        } catch (e: Exception) {
            result.error("INITIALIZATION_ERROR", e.message, null)
        }
    }

    private fun handleLoad(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val mediaItem = call.argument<Map<String, Any>>("mediaItem")
            
            if (playerId != null && mediaItem != null) {
                playerManager.loadMediaItem(playerId, mediaItem)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and media item are required", null)
            }
        } catch (e: Exception) {
            result.error("LOAD_ERROR", e.message, null)
        }
    }

    private fun handleSetPlaylist(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val playlist = call.argument<Map<String, Any>>("playlist")
            val startIndex = call.argument<Int>("startIndex") ?: 0
            
            if (playerId != null && playlist != null) {
                playerManager.setPlaylist(playerId, playlist, startIndex)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and playlist are required", null)
            }
        } catch (e: Exception) {
            result.error("PLAYLIST_ERROR", e.message, null)
        }
    }

    private fun handlePlay(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            if (playerId != null) {
                playerManager.play(playerId)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID is required", null)
            }
        } catch (e: Exception) {
            result.error("PLAY_ERROR", e.message, null)
        }
    }

    private fun handlePause(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            if (playerId != null) {
                playerManager.pause(playerId)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID is required", null)
            }
        } catch (e: Exception) {
            result.error("PAUSE_ERROR", e.message, null)
        }
    }

    private fun handleStop(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            if (playerId != null) {
                playerManager.stop(playerId)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID is required", null)
            }
        } catch (e: Exception) {
            result.error("STOP_ERROR", e.message, null)
        }
    }

    private fun handleSeekTo(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val position = call.argument<Int>("position")
            
            if (playerId != null && position != null) {
                playerManager.seekTo(playerId, position)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and position are required", null)
            }
        } catch (e: Exception) {
            result.error("SEEK_ERROR", e.message, null)
        }
    }

    private fun handleSetVolume(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val volume = call.argument<Double>("volume")
            
            if (playerId != null && volume != null) {
                playerManager.setVolume(playerId, volume.toFloat())
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and volume are required", null)
            }
        } catch (e: Exception) {
            result.error("VOLUME_ERROR", e.message, null)
        }
    }

    private fun handleSetSpeed(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val speed = call.argument<Double>("speed")
            
            if (playerId != null && speed != null) {
                playerManager.setPlaybackSpeed(playerId, speed.toFloat())
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and speed are required", null)
            }
        } catch (e: Exception) {
            result.error("SPEED_ERROR", e.message, null)
        }
    }

    private fun handleSetMuted(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val muted = call.argument<Boolean>("muted")
            
            if (playerId != null && muted != null) {
                playerManager.setMuted(playerId, muted)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and muted state are required", null)
            }
        } catch (e: Exception) {
            result.error("MUTE_ERROR", e.message, null)
        }
    }

    private fun handleSetBoxFit(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val boxFit = call.argument<String>("boxFit")
            
            if (playerId != null && boxFit != null) {
                playerManager.setBoxFit(playerId, boxFit)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and box fit are required", null)
            }
        } catch (e: Exception) {
            result.error("BOXFIT_ERROR", e.message, null)
        }
    }

    private fun handleSetSubtitleTrack(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val subtitleTrack = call.argument<Map<String, Any>?>("subtitleTrack")
            
            if (playerId != null) {
                playerManager.setSubtitleTrack(playerId, subtitleTrack)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID is required", null)
            }
        } catch (e: Exception) {
            result.error("SUBTITLE_ERROR", e.message, null)
        }
    }

    private fun handleSetQualityTrack(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val qualityTrack = call.argument<Map<String, Any>>("qualityTrack")
            
            if (playerId != null && qualityTrack != null) {
                playerManager.setQualityTrack(playerId, qualityTrack)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and quality track are required", null)
            }
        } catch (e: Exception) {
            result.error("QUALITY_ERROR", e.message, null)
        }
    }

    private fun handleSetAudioTrack(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val audioTrack = call.argument<Map<String, Any>>("audioTrack")
            
            if (playerId != null && audioTrack != null) {
                playerManager.setAudioTrack(playerId, audioTrack)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and audio track are required", null)
            }
        } catch (e: Exception) {
            result.error("AUDIO_ERROR", e.message, null)
        }
    }

    private fun handleEnableAutoQuality(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            
            if (playerId != null) {
                playerManager.enableAutoQuality(playerId)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID is required", null)
            }
        } catch (e: Exception) {
            result.error("AUTO_QUALITY_ERROR", e.message, null)
        }
    }

    private fun handleSkipToIndex(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val index = call.argument<Int>("index")
            
            if (playerId != null && index != null) {
                playerManager.skipToIndex(playerId, index)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and index are required", null)
            }
        } catch (e: Exception) {
            result.error("SKIP_ERROR", e.message, null)
        }
    }

    private fun handleUpdateConfig(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            val config = call.argument<Map<String, Any>>("config")
            
            if (playerId != null && config != null) {
                playerManager.updateConfig(playerId, config)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID and config are required", null)
            }
        } catch (e: Exception) {
            result.error("CONFIG_ERROR", e.message, null)
        }
    }

    private fun handleDispose(call: MethodCall, result: Result) {
        try {
            val playerId = call.argument<String>("playerId")
            if (playerId != null) {
                playerManager.disposePlayer(playerId)
                result.success(null)
            } else {
                result.error("INVALID_ARGUMENT", "Player ID is required", null)
            }
        } catch (e: Exception) {
            result.error("DISPOSE_ERROR", e.message, null)
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        playerManager.dispose()
    }
}
